src/StringTools/WeakStringCache.Concurrent.cs

```diff
@@ -0,0 +1,125 @@
+
+        public WeakStringCache()
+        {
+            _stringsByHashCode = new ConcurrentDictionary<int, StringWeakHandle>(Environment.ProcessorCount, _initialCapacity);
+        }
+
+        /// <summary>
+        /// Main entrypoint of this cache. Tries to look up a string that matches the given internable. If it succeeds, returns
+        /// the string and sets cacheHit to true. If the string is not found, calls ExpensiveConvertToString on the internable,
+        /// adds the resulting string to the cache, and returns it, setting cacheHit to false.
+        /// </summary>
+        /// <param name="internable">The internable describing the string we're looking for.</param>
+        /// <param name="cacheHit">true if match found in cache, false otherwise.</param>
+        /// <returns>A string matching the given internable.</returns>
+        public string GetOrCreateEntry(ref InternableString internable, out bool cacheHit)
+        {
+            int hashCode = internable.GetHashCode();
+
+            StringWeakHandle handle;
+            string result;
+
+            // Get the existing handle from the cache and lock it while we're dereferencing it to prevent a race with the Scavenge
+            // method running on another thread and freeing the handle from underneath us.
+            if (_stringsByHashCode.TryGetValue(hashCode, out handle))
+            {
+                lock (handle)
+                {
+                    result = handle.GetString(ref internable);
+                    if (result != null)
+                    {
+                        cacheHit = true;
+                        return result;
+                    }
+
+                    // We have the handle but it's not referencing the right string - create the right string and store it in the handle.
+                    result = internable.ExpensiveConvertToString();
+                    handle.SetString(result);
+
+                    cacheHit = false;
+                    return result;
+                }
+            }
+
+            // We don't have the handle in the cache - create the right string, store it in the handle, and add the handle to the cache.
+            result = internable.ExpensiveConvertToString();
+
+            handle = new StringWeakHandle();
+            handle.SetString(result);
+            _stringsByHashCode.TryAdd(hashCode, handle);
+
+            // Remove unused handles if our heuristic indicates that it would be productive.
+            int scavengeThreshold = _scavengeThreshold;
+            if (_stringsByHashCode.Count >= scavengeThreshold)
+            {
+                // Before we start scavenging set _scavengeThreshold to a high value to effectively lock other threads from
+                // running Scavenge at the same time.
+                if (Interlocked.CompareExchange(ref _scavengeThreshold, int.MaxValue, scavengeThreshold) == scavengeThreshold)
```

Comment:
I don't think I understand why this works, partially because I would've thought that since the scavengerThreshold is the same as for other threads, if two threads separately decide to run Scavenge, they would both try to set the lock to scavengeThreshold. Am I misunderstanding how Interlocked.CompareExchange works? Or would it be helpful to add a small random number to the scavengeThreshold?

Reply:
If two threads get to this line because the size of the dictionary has exceeded `scavengeThreshold`, they both try to replace `scavengeThreshold` with `int.MaxValue` and the `Interlocked.CompareExchange` call ensures that only one will succeed. That's the one to proceed and run scavenging while the other one will simply skip and not worry about it. When scavenging is in progress, the condition `_stringsByHashCode.Count >= scavengeThreshold` will not hold for anyone and they won't even need the interlocked operation (which is not that cheap either).

It's basically a lock-free implementation of mutual exclusion. We don't want to run scavenging under a lock because we don't want other threads to stop and wait for it to be finished. We just need a lock to "elect" the scavenger thread on contention and this is most efficiently done with this one interlocked operation.

Reply:
I had to read that several times, but I think I get it now. Thanks for the explanation!